In [1]:
from dotenv import load_dotenv

# 加载环境变量
load_dotenv()

True

## 创建文档

数据来源：

CRUD-RAG：大型语言模型检索增强生成的综合中文基准

[IAAR-Shanghai/CRUD-RAG](https://github.com/IAAR-Shanghai/CRUD_RAG/tree/main)



In [2]:
from langchain_community.document_loaders import TextLoader

In [3]:
loaders = [
    TextLoader("data/news1.txt", encoding="utf8"), 
    TextLoader("data/news2.txt", encoding="utf8"),
]
docs = []
for loader in loaders:
    docs.extend(loader.load())

print(len(docs))

2


In [4]:
docs

[Document(page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023 年第二季度的研发费用从 2022 年第二季度的 40 亿元人民币增长了 1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 60 多亿元建设的“京东青年城”也正式启动，未来将提供近 4000 套可拎包入住的员工公寓。据IT之家此前报道，京东物流第二季度经调整后净利润 8.3 亿元，同比增长 288%；营收 410 亿元，同比增长 31.2%。广告声明：文内含有的对外跳转链接（包括不限于超链接、二维码、口令等形式），用于传递更多信息，节省甄选时间，结果仅供参考，IT之家所有文章均包含本声明。', metadata={'source': 'data/news1.txt'}),
 Document(page_content='2023-08-16 21:12:47，正文：IT之家 8 月 16 日消息，腾讯控股今日发布了最新的 Q2 财报。财报显示，腾讯二季度研发开支达 160 亿元，自 2018 年至今累计投入超过 

In [5]:
print(docs[0])

page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023 年第二季度的研发费用从 2022 年第二季度的 40 亿元人民币增长了 1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 60 多亿元建设的“京东青年城”也正式启动，未来将提供近 4000 套可拎包入住的员工公寓。据IT之家此前报道，京东物流第二季度经调整后净利润 8.3 亿元，同比增长 288%；营收 410 亿元，同比增长 31.2%。广告声明：文内含有的对外跳转链接（包括不限于超链接、二维码、口令等形式），用于传递更多信息，节省甄选时间，结果仅供参考，IT之家所有文章均包含本声明。' metadata={'source': 'data/news1.txt'}


In [6]:
print(len(docs[0].page_content))

811


> ### 京东集团 2023 年第二季度财报要点：
>
> #### 营收与利润
>
> 1. **总营收**：2879 亿元人民币，同比增长 7.6%。
> 2. **净服务收入**：541 亿元人民币，同比增长 30.1%。
> 3. **运营收入**：83 亿元人民币，同比增加 45 亿元人民币。
> 4. **非 GAAP 运营收入**：87 亿元人民币，同比增加 29 亿元人民币。
> 5. **净收入**：归属于普通股股东的净收入为 66 亿元人民币，同比增加 22 亿元人民币。
> 6. **非 GAAP 净收入**：86 亿元人民币，同比增长 31.9%。
>
> #### 利润率与研发费用
>
> 1. **营业利润率**：京东零售未分配项目前的营业利润率为 3.2%，去年同期为 3.4%。
> 2. **研发费用**：41 亿元人民币，同比增长 1.1%。
> 3. **研发费用占比**：占净收入的比例为 1.4%，去年同期为 1.5%。
>
> #### 员工与项目
>
> 1. **员工总数**：截至二季度末，京东体系员工总数超 56 万。
> 2. **京东青年城**：预计投入 60 多亿元建设，提供近 4000 套可拎包入住的员工公寓。
>
> #### 京东物流
>
> 1. **净利润**：经调整后净利润为 8.3 亿元人民币，同比增长 288%。
> 2. **营收**：营收 410 亿元人民币，同比增长 31.2%。

In [7]:
print(docs[1])

page_content='2023-08-16 21:12:47，正文：IT之家 8 月 16 日消息，腾讯控股今日发布了最新的 Q2 财报。财报显示，腾讯二季度研发开支达 160 亿元，自 2018 年至今累计投入超过 2300 亿元。据称，腾讯研发投入和专利数量继续保持高位，行业大模型加速落地驱动产业升级，科技底色更足。在腾讯财报的电话会议中，腾讯高管表示，基础大模型正在顺利推进中，今年下半年会有升级。腾讯表示，AI 是一个增长的催化剂，可以增长效率、还包括对用户服务的质量。同时 AI 还可以更好地促进应用对用户的针对性，还能提升成本效率。截至 6 月，腾讯在全球主要国家和地区专利申请公开总数超过 6.6 万件，专利授权数超过 3.3 万件，主要集中在人工智能、云技术、大出行、即时通讯等前沿领域。据IT之家此前报道，BAT 中的其他两家都已经拿出了各自的通用大模型，并开启一系列企业合作以及个人用户测试。马化腾此前曾对此表示“腾讯也在埋头研发，并不急于把半成品拿出来展示”。值得一提的是，36 氪此前报道称，腾讯自研的“腾讯混元大模型”已经进入应用内测阶段，多位腾讯员工表示，近期收到了内部邮件邀请，可以通过网页或者小程序体验混元大模型。据了解，腾讯内部多个业务已经接入混元大模型进行测试。今年 2 月份，36 氪就称腾讯针对类 ChatGPT 对话式产品已成立“混元助手 (HunyuanAide)”项目组。报道称，混元助手项目组的一号位张正友是腾讯史上最高职级拥有者，早年曾在微软研究院就职，带领视觉团队从事大量学术研究，贡献颇丰。张正友于 2018 年加入腾讯，创建了腾讯机器人实验室 Robotics X，致力推进人机协作的下一代机器人研究。2021 年 1 月 8 日，腾讯宣布张正友成为腾讯首位 17 级研究员 / 杰出科学家。“腾讯混元大模型”的实力究竟如何，还需要等官方正式公开并测试才能得知了。相关阅读：《腾讯首次披露行业大模型进展：不追求参数高，希望成本预算可控》广告声明：文内含有的对外跳转链接（包括不限于超链接、二维码、口令等形式），用于传递更多信息，节省甄选时间，结果仅供参考，IT之家所有文章均包含本声明。' metadata={'source': 'data/news2.txt'}


In [8]:
print(len(docs[1].page_content))

911


> ### 腾讯2023年Q2财报要点
>
> #### 财务数据
>
> - 二季度研发开支达160亿元。
> - 2018年至今累计研发投入超过2300亿元。
>
> #### 研发与创新
>
> - 研发投入和专利数量持续保持高位。
> - 行业大模型加速落地，驱动产业升级。
> - 基础大模型顺利推进，下半年将有升级。
>
> #### 人工智能
>
> - AI被视为增长催化剂，提升效率和用户服务质量。
> - AI促进应用对用户的针对性，提高成本效率。
>
> #### 专利数据
>
> - 截至6月，全球专利申请公开总数超过6.6万件。
> - 专利授权数超过3.3万件，集中在AI、云技术、大出行、即时通讯等领域。
>
> #### 腾讯混元大模型
>
> - 腾讯混元大模型已进入应用内测阶段。
> - 多个业务已接入混元大模型进行测试。
> - 内部成立“混元助手 (HunyuanAide)”项目组，推进类ChatGPT对话式产品。
> - 项目组由张正友带领，致力于下一代机器人研究。
>
> #### 行业内竞争
>
> - BAT中的其他两家公司已推出通用大模型并开展合作及测试。
> - 马化腾表示腾讯埋头研发，不急于展示半成品。

## 分步实现

### 创建文档分割器

In [9]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

# 这个文本拆分器用于创建子文档
child_splitter = RecursiveCharacterTextSplitter(chunk_size=200, chunk_overlap=0, add_start_index=True)

# 这个文本拆分器用于创建父文档
parent_splitter = RecursiveCharacterTextSplitter(chunk_size=500, chunk_overlap=100, add_start_index=True)

#### 分割文档

In [10]:
parent_docs = parent_splitter.split_documents(docs)
print(len(parent_docs))
parent_docs

5


[Document(page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023', metadata={'source': 'data/news1.txt', 'start_index': 0}),
 Document(page_content='44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023 年第二季度的研发费用从 2022 年第二季度的 40 亿元人民币增长了 1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 60 多亿元建设的“京东青年城”也正式启动，未来将提供近 4000 套可拎包入住的员工公寓。据IT之家此前报道，京东物流第二季度经调整后净利润 8.3 亿元，同比增长 288%；营收 410 亿元，同比增长 31.2%。广告声明：文内含有的对外跳转链接（包括不限于超链接、二维码、口令等形式），用于传递更多信息，节省甄选时间，结果仅供

In [12]:
import uuid

parent_doc_ids = [str(uuid.uuid4()) for _ in parent_docs]
parent_doc_ids

['72f10c51-abc7-478b-91ce-231576ed8685',
 'bd20972a-006e-4a90-8d20-eedd5ba61d88',
 'de41562a-82a2-4a80-a3d9-9723fe8bd295',
 'b490c41a-9434-41ee-bb94-2e1a23a4dda4',
 'c4cf2463-4b6b-41cb-bc90-1392426d61e7']

In [13]:
all_sub_docs = []
full_docs = []
for i, doc in enumerate(parent_docs):
    _id = parent_doc_ids[i]
    sub_docs = child_splitter.split_documents([doc])
    for _doc in sub_docs:
        _doc.metadata["doc_id"] = _id
    all_sub_docs.extend(sub_docs)
    full_docs.append((_id, doc))

In [14]:
print(len(all_sub_docs))

14


In [15]:
all_sub_docs

[Document(page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为', metadata={'source': 'data/news1.txt', 'start_index': 0, 'doc_id': '72f10c51-abc7-478b-91ce-231576ed8685'}),
 Document(page_content='83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为', metadata={'source': 'data/news1.txt', 'start_index': 200, 'doc_id': '72f10c51-abc7-478b-91ce-231576ed8685'}),
 Document(page_content='44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023', metadata={'source': 'data/news1.txt', 'start_index': 399, 'doc_id': '72f10c51-abc7-478b-91ce-231576ed8685'}),
 Document(page_content='44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元

In [16]:
print(len(full_docs))

5


In [17]:
full_docs

[('72f10c51-abc7-478b-91ce-231576ed8685',
  Document(page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023', metadata={'source': 'data/news1.txt', 'start_index': 0})),
 ('bd20972a-006e-4a90-8d20-eedd5ba61d88',
  Document(page_content='44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023 年第二季度的研发费用从 2022 年第二季度的 40 亿元人民币增长了 1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 60 多亿元建设的“京东青年城”也正式启动，未来将提供近 4000 套可拎包入住的员工公寓。据IT之家此前报道，京东物流第二季度经调整后净利润 8.3 亿元，同

### 创建父文档存储

In [18]:
from langchain.storage import InMemoryStore

# 父文档的存储层
parent_store = InMemoryStore()

In [19]:
parent_store

In [20]:
# 查看store中存储的父文档的ID
list(parent_store.yield_keys())

[]

In [21]:
parent_store.mset(full_docs)

In [22]:
# 查看store中存储的父文档的ID
list(parent_store.yield_keys())

['72f10c51-abc7-478b-91ce-231576ed8685',
 'bd20972a-006e-4a90-8d20-eedd5ba61d88',
 'de41562a-82a2-4a80-a3d9-9723fe8bd295',
 'b490c41a-9434-41ee-bb94-2e1a23a4dda4',
 'c4cf2463-4b6b-41cb-bc90-1392426d61e7']

### 创建子文档存储 —— 向量存储

In [23]:
from langchain_openai import OpenAIEmbeddings

embeddings = OpenAIEmbeddings()

In [24]:
# from langchain_community.vectorstores import Qdrant

# qdrant = Qdrant.from_documents(
#     docs,
#     embeddings,
#     location=":memory:",  # 本地模式，不使用 Qdrant 服务器，只使用内存存储
#     collection_name="my_documents",  # 集合名称
# )

# qdrant = Qdrant.from_documents(
#     docs,
#     embeddings,
#     path="./data/local_qdrant", # 本地模式，不使用 Qdrant 服务器，将向量存储在磁盘指定位置
#     collection_name="my_documents",  # 集合名称
# )

# from qdrant_client import QdrantClient

# qdrant = Qdrant(
#     client=QdrantClient("http://localhost:6333"), 
#     collection_name="my_documents", 
#     embeddings
# )

In [25]:
from langchain_community.vectorstores import Qdrant
from qdrant_client import QdrantClient

collection_name = "full_documents"

qdrant = Qdrant(
    client=QdrantClient(url="http://localhost:6333"), 
    collection_name=collection_name, 
    embeddings=embeddings
)

D:\Software\anaconda3\envs\py310_LLMAppCourse2\lib\site-packages\langchain_core\_api\deprecation.py:119: LangChainDeprecationWarning: The class `Qdrant` was deprecated in LangChain 0.0.37 and will be removed in 0.3.0. An updated version of the class exists in the langchain-qdrant package and should be used instead. To use it run `pip install -U langchain-qdrant` and import as `from langchain_qdrant import Qdrant`.
  warn_deprecated(


In [26]:
from qdrant_client.http.models import Distance, VectorParams

# 向量维度
size = 1536

# 若集合已存在，删除该集合
if qdrant.client.collection_exists(collection_name):
    print(f"集合已存在 | collection_name: {collection_name}")
    qdrant.client.delete_collection(collection_name)
    print(f"集合已删除 | collection_name: {collection_name}")
    
# 创建新集合
if qdrant.client.create_collection(
    collection_name=collection_name,
    vectors_config=VectorParams(
                size=size, distance=Distance.COSINE)
):
    print(f"创建集合成功 | collection_name: {collection_name}")
else:
    print(f"创建集合失败 | collection_name: {collection_name}")

集合已存在 | collection_name: full_documents
集合已删除 | collection_name: full_documents
创建集合成功 | collection_name: full_documents


In [27]:
qdrant.add_documents(all_sub_docs)

['399331a39926400aa595eb59f1fb7265',
 '7d3724c6cce34cd19fb5f5540864c9f4',
 'ccd746db66984914904d620955809fa0',
 'f9e6900c1e86439080b78ecc1c2290f2',
 'daf40fa47f314b1dbbc1f333d6ee4f95',
 '42e3bccd403642c69452289e4cbc9a7f',
 '13d0caea9ba847228ec88646c21a0274',
 'cc26f6e4753145329467a8d5596ac762',
 'ad2711d86c8045df8fd1b0c0ecf26529',
 'f762f88c861b486c99eca7489abdf345',
 'fb85e65b362b43b0a86417792889f9d8',
 'f409fc83d3c6424e90f065c831f5286d',
 '7ebec14ef7b64c1abcbf2630a643e2c6',
 'e8726afcc85c4ca89c4c06c4b8dd637f']

### 检索

In [28]:
query = "京东物流营收如何？"

In [29]:
# 搜索与 query 相似度较高的子文档
retrieved_sub_docs = qdrant.similarity_search(query)

print(len(retrieved_sub_docs))

for doc in retrieved_sub_docs:
    print(doc)
    print()

# print(retrieved_sub_docs[0].page_content)

4
page_content='1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 60 多亿元建设的“京东青年城”也正式启动，未来将提供近 4000 套可拎包入住的员工公寓。据IT之家此前报道，京东物流第二季度经调整后净利润 8.3 亿元，同比增长 288%；营收 410 亿元，同比增长' metadata={'doc_id': 'bd20972a-006e-4a90-8d20-eedd5ba61d88', 'source': 'data/news1.txt', 'start_index': 195, '_id': 'daf40fa4-7f31-4b1d-bbc1-f333d6ee4f95', '_collection_name': 'full_documents'}

page_content='83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为' metadata={'doc_id': '72f10c51-abc7-478b-91ce-231576ed8685', 'source': 'data/news1.txt', 'start_index': 200, '_id': '7d3724c6-cce3-4cd1-9fb5-f5540864c9f4', '_collection_name': 'full_documents'}

page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为' metadata={'doc_id': '72f10c51-

In [30]:
# 返回父文档ID
retrieved_parent_ids = []
for d in retrieved_sub_docs:
    if "doc_id" in d.metadata and d.metadata["doc_id"] not in retrieved_parent_ids:
        retrieved_parent_ids.append(d.metadata["doc_id"])
retrieved_parent_ids

['bd20972a-006e-4a90-8d20-eedd5ba61d88',
 '72f10c51-abc7-478b-91ce-231576ed8685']

In [31]:
retrieved_parent_docs = parent_store.mget(retrieved_parent_ids)
print(len(retrieved_parent_docs))

retrieved_parent_docs

2


[Document(page_content='44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023 年第二季度的研发费用从 2022 年第二季度的 40 亿元人民币增长了 1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 60 多亿元建设的“京东青年城”也正式启动，未来将提供近 4000 套可拎包入住的员工公寓。据IT之家此前报道，京东物流第二季度经调整后净利润 8.3 亿元，同比增长 288%；营收 410 亿元，同比增长 31.2%。广告声明：文内含有的对外跳转链接（包括不限于超链接、二维码、口令等形式），用于传递更多信息，节省甄选时间，结果仅供参考，IT之家所有文章均包含本声明。', metadata={'source': 'data/news1.txt', 'start_index': 399}),
 Document(page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比

## 使用ParentDocumentRetriever实现

### 创建文档分割器

In [33]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

In [34]:
# 这个文本拆分器用于创建子文档
# 应该创建比父文档更小的文档
child_splitter = RecursiveCharacterTextSplitter(chunk_size=200, chunk_overlap=0, add_start_index=True)

# 这个文本拆分器用于创建父文档
parent_splitter = RecursiveCharacterTextSplitter(chunk_size=500, chunk_overlap=100, add_start_index=True)

### 创建父文档存储

In [35]:
from langchain.storage import InMemoryStore

In [36]:
# 父文档的存储层
store = InMemoryStore()

### 创建子文档存储 —— 向量存储

In [37]:
from langchain_openai import OpenAIEmbeddings

embeddings = OpenAIEmbeddings()

In [38]:
from langchain_community.vectorstores import Qdrant
from qdrant_client import QdrantClient

collection_name = "full_documents_1"

qdrant = Qdrant(
    client=QdrantClient(url="http://localhost:6333"), 
    collection_name=collection_name, 
    embeddings=embeddings
)

In [39]:
from qdrant_client.http.models import Distance, VectorParams

# 向量维度
size = 1536

# 若集合已存在，删除该集合
if qdrant.client.collection_exists(collection_name):
    print(f"集合已存在 | collection_name: {collection_name}")
    qdrant.client.delete_collection(collection_name)
    print(f"集合已删除 | collection_name: {collection_name}")
    
# 创建新集合
if qdrant.client.create_collection(
    collection_name=collection_name,
    vectors_config=VectorParams(
                size=size, distance=Distance.COSINE)
):
    print(f"创建集合成功 | collection_name: {collection_name}")
else:
    print(f"创建集合失败 | collection_name: {collection_name}")

集合已存在 | collection_name: full_documents_1
集合已删除 | collection_name: full_documents_1
创建集合成功 | collection_name: full_documents_1


### 创建检索器

In [40]:
from langchain.retrievers import ParentDocumentRetriever

In [41]:
# 初始化检索器
retriever = ParentDocumentRetriever(
    vectorstore=qdrant,
    docstore=store,
    child_splitter=child_splitter,
    parent_splitter=parent_splitter,
)

### 存储文档

In [42]:
store

In [43]:
# 查看store中存储的父文档的ID
list(store.yield_keys())

[]

In [44]:
docs

[Document(page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023 年第二季度的研发费用从 2022 年第二季度的 40 亿元人民币增长了 1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 60 多亿元建设的“京东青年城”也正式启动，未来将提供近 4000 套可拎包入住的员工公寓。据IT之家此前报道，京东物流第二季度经调整后净利润 8.3 亿元，同比增长 288%；营收 410 亿元，同比增长 31.2%。广告声明：文内含有的对外跳转链接（包括不限于超链接、二维码、口令等形式），用于传递更多信息，节省甄选时间，结果仅供参考，IT之家所有文章均包含本声明。', metadata={'source': 'data/news1.txt'}),
 Document(page_content='2023-08-16 21:12:47，正文：IT之家 8 月 16 日消息，腾讯控股今日发布了最新的 Q2 财报。财报显示，腾讯二季度研发开支达 160 亿元，自 2018 年至今累计投入超过 

In [45]:
# 将文档添加到文档存储和向量存储中
# ids:文档的ID列表。如果未提供，则生成UUID。
retriever.add_documents(docs, ids=None)

In [46]:
# 查看store中存储的父文档的ID
list(store.yield_keys())

['9a2d90f9-58aa-49fb-9caa-6ea4cadf58af',
 'd9e3bc2f-9717-4abf-b94e-b52f4979b949',
 'a1229ed4-be59-4faa-a28e-f0de6ec151ca',
 '12c10716-3375-4f56-bef7-e57deee090ce',
 'bc40af4b-2d0e-4b6e-a41b-d6227d15e3f2']

### 检索

检索器返回的是父文档（大chunk），即子文档（小chunk）所属的父文档（大chunk）

In [47]:
query = "京东物流营收情况如何？"

In [48]:
retrieved_docs = retriever.invoke(query)

print(len(retrieved_docs))
print(retrieved_docs)

2
[Document(page_content='44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同比增长 31.9% 超出市场预期。2023 年第二季度的研发费用从 2022 年第二季度的 40 亿元人民币增长了 1.1%，达到 41 亿元人民币（6 亿美元）。2023 年第二季度，研发费用占净收入的比例为 1.4%，而去年同期为 1.5%。截至二季度末，京东体系员工总数超 56 万，预计投入 60 多亿元建设的“京东青年城”也正式启动，未来将提供近 4000 套可拎包入住的员工公寓。据IT之家此前报道，京东物流第二季度经调整后净利润 8.3 亿元，同比增长 288%；营收 410 亿元，同比增长 31.2%。广告声明：文内含有的对外跳转链接（包括不限于超链接、二维码、口令等形式），用于传递更多信息，节省甄选时间，结果仅供参考，IT之家所有文章均包含本声明。', metadata={'source': 'data/news1.txt', 'start_index': 399}), Document(page_content='2023-08-16 18:22:06，正文：IT之家 8 月 16 日消息，京东集团今日发布 Q2 财报，2023 年第二季度的营收为 2879 亿元人民币（397 亿美元），比 2022 年第二季度同比增长 7.6%。2023 年第二季度的净服务收入为 541 亿元人民币（75 亿美元），比 2022 年第二季增长 30.1%，官方称“核心指标全面超出预期”。2023 年第二季度的运营收入为 83 亿元人民币（11 亿美元），而去年同期为 38 亿元人民币。2023 年第二季度，非 GAAP 运营收入为 87 亿元人民币（12 亿美元），而 2022 年第二季度为 58 亿元人民币。京东零售 2023 年第二季度未分配项目前的营业利润率为 3.2%，而 2022 年第二季度为 3.4%。2023 年第二季度归属于公司普通股股东的净收入为 66 亿元人民币（9 亿美元），而去年同期为 44 亿元人民币。2023 年第二季度，归属于公司普通股股东的非 GAAP 净收入为 86 亿元人民币（12 亿美元），而去年同期为 65 亿元人民币，同

In [49]:
query = "腾讯开发了什么大模型？"

In [50]:
retrieved_docs = retriever.invoke(query)

print(len(retrieved_docs))
print(retrieved_docs)

3
[Document(page_content='2018 年加入腾讯，创建了腾讯机器人实验室 Robotics X，致力推进人机协作的下一代机器人研究。2021 年 1 月 8 日，腾讯宣布张正友成为腾讯首位 17 级研究员 / 杰出科学家。“腾讯混元大模型”的实力究竟如何，还需要等官方正式公开并测试才能得知了。相关阅读：《腾讯首次披露行业大模型进展：不追求参数高，希望成本预算可控》广告声明：文内含有的对外跳转链接（包括不限于超链接、二维码、口令等形式），用于传递更多信息，节省甄选时间，结果仅供参考，IT之家所有文章均包含本声明。', metadata={'source': 'data/news2.txt', 'start_index': 667}), Document(page_content='中的其他两家都已经拿出了各自的通用大模型，并开启一系列企业合作以及个人用户测试。马化腾此前曾对此表示“腾讯也在埋头研发，并不急于把半成品拿出来展示”。值得一提的是，36 氪此前报道称，腾讯自研的“腾讯混元大模型”已经进入应用内测阶段，多位腾讯员工表示，近期收到了内部邮件邀请，可以通过网页或者小程序体验混元大模型。据了解，腾讯内部多个业务已经接入混元大模型进行测试。今年 2 月份，36 氪就称腾讯针对类 ChatGPT 对话式产品已成立“混元助手 (HunyuanAide)”项目组。报道称，混元助手项目组的一号位张正友是腾讯史上最高职级拥有者，早年曾在微软研究院就职，带领视觉团队从事大量学术研究，贡献颇丰。张正友于 2018 年加入腾讯，创建了腾讯机器人实验室 Robotics X，致力推进人机协作的下一代机器人研究。2021 年 1 月 8 日，腾讯宣布张正友成为腾讯首位 17 级研究员 /', metadata={'source': 'data/news2.txt', 'start_index': 356}), Document(page_content='2023-08-16 21:12:47，正文：IT之家 8 月 16 日消息，腾讯控股今日发布了最新的 Q2 财报。财报显示，腾讯二季度研发开支达 160 亿元，自 2018 年至今累计投入超过 2300 亿元。据称，腾讯研发投入和专利数量继续保持高位，行业大模型加速落地驱动产业升级，科技底色更足。在腾讯财报的电

In [51]:
# 搜索与 query 相似度较高的子文档
qdrant.similarity_search(query)

[Document(page_content='杰出科学家。“腾讯混元大模型”的实力究竟如何，还需要等官方正式公开并测试才能得知了。相关阅读：《腾讯首次披露行业大模型进展：不追求参数高，希望成本预算可控》广告声明：文内含有的对外跳转链接（包括不限于超链接、二维码、口令等形式），用于传递更多信息，节省甄选时间，结果仅供参考，IT之家所有文章均包含本声明。', metadata={'doc_id': 'bc40af4b-2d0e-4b6e-a41b-d6227d15e3f2', 'source': 'data/news2.txt', 'start_index': 91, '_id': '1cfa4d95-644b-4397-966f-4685caea42b6', '_collection_name': 'full_documents_1'}),
 Document(page_content='中的其他两家都已经拿出了各自的通用大模型，并开启一系列企业合作以及个人用户测试。马化腾此前曾对此表示“腾讯也在埋头研发，并不急于把半成品拿出来展示”。值得一提的是，36 氪此前报道称，腾讯自研的“腾讯混元大模型”已经进入应用内测阶段，多位腾讯员工表示，近期收到了内部邮件邀请，可以通过网页或者小程序体验混元大模型。据了解，腾讯内部多个业务已经接入混元大模型进行测试。今年 2 月份，36', metadata={'doc_id': '12c10716-3375-4f56-bef7-e57deee090ce', 'source': 'data/news2.txt', 'start_index': 0, '_id': '46c19e1a-9de1-42e7-bbc9-2a94b4403fd0', '_collection_name': 'full_documents_1'}),
 Document(page_content='中的其他两家都已经拿出了各自的通用大模型，并开启一系列企业合作以及个人用户测试。马化腾此前曾对此表示“腾讯也在埋头研发，并不急于把半成品拿出来展示”。值得一提的是，36', metadata={'doc_id': 'a1229ed4-be59-4faa-a28e-f0de6ec151ca', 'source': 'data/news2.txt', 'start_ind